### PR TITLE
Make channels unidirectional

### PR DESCRIPTION
**Pull request by daviddrysdale**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -126,19 +126,21 @@ Each Oak Module must expose the following **exported functions** as
...
+
+At each invocation, the following channel halfs are implicitly available to the Oak
 Node (see [/oak/server/oak_node.h](oak/server/oak_node.h)):
 
--   `logging` (handle: 1)
--   `grpc` (handle: 2)
--   `grpc_method` (handle: 3)
+-   `logging` (handle: 1, send)
+-   `grpc_method` (handle: 2, receive)
+-   `grpc_in` (handle: 3, send)
```
Comment (tiziano88) [inline README.md]: Are these backwards? I would imagine apps would read from `grpc_in` and write to `grpc_out`?

File: oak/server/buffer_channel.h [hunk 2]
```diff
@@ -22,38 +22,45 @@
...
+// buffer.  The channel does not own the provided buffer, so the caller must ensure
+// its lifetime is longer than that of the channel.
+class WriteBufferChannelHalf final : public ChannelHalf {
+ public:
+  WriteBufferChannelHalf(std::vector<char>* data) : data_(data) {}
 
   uint32_t Write(absl::Span<const char> data) override {
     LOG(INFO) << "Writing to channel: " << data.size() << " bytes";
-    if (data_output_ == nullptr) {
+    if (data_ == nullptr) {
```
Comment (tiziano88) [inline oak/server/buffer_channel.h]: We should probably check for this in the constructor, now that the two halves are distinct there is no point in creating this half with a null data.

File: oak/server/oak_node.h [hunk 3]
```diff
@@ -82,6 +80,20 @@ class OakNode final : public Node::Service {
   const std::string module_hash_sha_256_;
 };
 
+// RAII class to add a mapping from a handle to a channel half for the duration of a scope.
```
Comment (tiziano88) [inline oak/server/oak_node.h]: Nice pattern!

File: README.md [hunk 4]
```diff
@@ -160,7 +162,7 @@ functions** as
 -   `channel_write: (i64, i32, i32) -> i32`: Writes data to the specified
     channel.
 
-    *   arg 0: Channel handle
+    *   arg 0: Handler to channel send half
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply README.md]: Fixed.

File: oak/server/oak_node.h [hunk 5]
```diff
@@ -1,4 +1,4 @@
-/*
+#/*
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply oak/server/oak_node.h]: Heh, I'm surprised that didn't cause problems &ndash; I guess it just got treated as an unknown preprocessor directive somehow.

Fixed.

File: oak/server/oak_node.h (hunk 3 above)
Comment (daviddrysdale) [reply oak/server/oak_node.h]: http://go/totw/41 ...

File: rust/oak/src/lib.rs [hunk 6]
```diff
@@ -39,34 +40,74 @@ mod wasm {
...
+impl ChannelPair {
+    pub fn new(in_handle: Handle, out_handle: Handle) -> ChannelPair {
+        ChannelPair {
+            receive: ReceiveChannelHalf::new(in_handle),
+            send: SendChannelHalf::new(out_handle),
+        }
+    }
+}
+
+impl Write for ChannelPair {
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply rust/oak/src/lib.rs]: Done.

File: oak/server/buffer_channel.h (hunk 2 above)
Comment (daviddrysdale) [reply oak/server/buffer_channel.h]: Added a check in the constructor, but we can't do more than log an error there without more major surgery (e.g. replacing the constructor with a factory method).

File: oak/server/oak_node.h (hunk 3 above)
Comment (tiziano88) [reply oak/server/oak_node.h]: You may like Rust hash map [entry API](https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html) then

**Pull request merged**